File: web/public/js/paginas/participacion.js
/* Tablero de participación (R2.1, R2.6).

   Tres preguntas, que son las tres que se le hacen a un panel:

     ¿Responde?                tasa de respuesta, global y por ola
     ¿Hace cuánto que no lo molesto?   tiempo desde el último contacto
     ¿Reparto parejo?          distribución de convocatorias

   La tercera es la que suele sorprender: si el decil más convocado se come
   la mayor parte de las convocatorias, la muestra no es del panel, es de ese
   pedacito del panel.
*/

import * as api from '../api.js';
import {
  $, $$, esc, encabezado, vacio, cargando, alerta, token, fechaCorta, hace,
} from '../ui.js';

let contexto = {};
let panelActual = null;

const TRAMOS_CONTACTO = {
  nunca: 'Nunca contactados',
  hasta_30_dias: 'Hasta 30 días',
  '31_a_90_dias': '31 a 90 días',
  '91_a_180_dias': '91 a 180 días',
  mas_de_180_dias: 'Más de 180 días',
};

const TRAMOS_CONVOCATORIAS = {
  0: 'Ninguna', 1: 'Una', 2: 'Dos', 3: 'Tres', 4: 'Cuatro', '5_o_mas': 'Cinco o más',
};

const pct = (n) => (n == null ? '—' : `${(n * 100).toFixed(1)} %`);

export async function render(main, ctx) {
  contexto = ctx;
  const { items: paneles } = await api.paneles.listar();
  if (!paneles.length) {
    main.innerHTML = encabezado('Participación', 'del panel', '')
      + vacio('No hay paneles todavía.', '📋');
    return;
  }
  panelActual = paneles.some((p) => p.id === panelActual) ? panelActual : paneles[0].id;

  main.innerHTML = encabezado('Participación', 'y fatiga',
    'Quién responde, a quién hace que no se contacta y cómo se reparten las convocatorias.') + `
    <div class="card"><div class="card-body">
      <div class="form-group" style="margin:0; max-width:340px"><label>Panel</label>
        <select class="fselect" id="panel">${paneles.map((p) =>
          `<option value="${p.id}" ${p.id === panelActual ? 'selected' : ''}>${esc(p.nombre)}</option>`
        ).join('')}</select></div>
    </div></div>
    <div id="cuerpo">${cargando()}</div>`;

  $('#panel').onchange = (e) => { panelActual = Number(e.target.value); cargar(); };
  await cargar();
}

async function cargar() {
  const caja = $('#cuerpo');
  caja.innerHTML = cargando();
  try {
    caja.innerHTML = pintar(await api.participacion.tablero(panelActual));
    $$('[data-panelista]', caja).forEach((fila) => {
      fila.onclick = () => contexto.irA('panelistas', { idPersona: fila.dataset.panelista });
    });
  } catch (error) {
    caja.innerHTML = alerta(error.message);
  }
}

function pintar(t) {
  const r = t.respuesta;
  const concentracion = t.convocatorias.concentracion_decil_superior;
  return `
    <div class="stat-grid">
      <div class="stat s-total"><label>Miembros activos</label><strong>${t.miembros}</strong></div>
      <div class="stat ${r.tasa_respuesta != null && r.tasa_respuesta >= 0.4 ? 's-ok' : 's-warn'}">
        <label>Tasa de respuesta</label><strong>${pct(r.tasa_respuesta)}</strong></div>
      <div class="stat s-free"><label>Convocatorias emitidas</label>
        <strong>${r.convocatorias_emitidas}</strong></div>
      <div class="stat ${r.miembros_nunca_convocados ? 's-warn' : 's-ok'}">
        <label>Nunca convocados</label><strong>${r.miembros_nunca_convocados}</strong></div>
    </div>

    ${concentracion != null && concentracion > 0.4 ? `<div class="alert alert-warn">
      El 10 % más convocado del panel se lleva el <strong>${pct(concentracion)}</strong>
      de las convocatorias. Cuando esa cifra sube, la muestra deja de ser del
      panel y pasa a ser de ese grupo.
    </div>` : ''}

    <div class="grid-2">
      <div class="card">
        <div class="card-header"><span class="card-header-title">Último contacto</span>
          <span class="small muted">mediana ${t.ultimo_contacto.mediana_dias ?? '—'} día(s)</span></div>
        <div class="card-body">${barras(
          t.ultimo_contacto.distribucion, TRAMOS_CONTACTO, t.miembros)}</div>
      </div>
      <div class="card">
        <div class="card-header"><span class="card-header-title">Convocatorias por miembro</span>
          <span class="small muted">promedio ${t.convocatorias.promedio_por_miembro ?? '—'}
            · máximo ${t.convocatorias.maximo}</span></div>
        <div class="card-body">${barras(
          t.convocatorias.distribucion, TRAMOS_CONVOCATORIAS, t.miembros)}</div>
      </div>
    </div>

    <div class="card">
      <div class="card-header"><span class="card-header-title">Olas</span></div>
      <div class="card-body tight">
        ${t.olas.length ? `<div class="table-wrap"><table>
          <thead><tr><th>Ola</th><th>Campo</th><th>Estado</th><th>Convocados</th>
            <th>Respondieron</th><th>Tasa</th><th>Calidad</th></tr></thead>
          <tbody>${t.olas.map((o) => `<tr>
            <td class="td-strong">${esc(o.nombre)}</td>
            <td class="small">${fechaCorta(o.fecha_campo)}</td>
            <td><span class="est est-${esc(o.estado)}">${esc(o.estado.replace('_', ' '))}</span></td>
            <td class="mono">${o.convocados}</td>
            <td class="mono">${o.respondieron}</td>
            <td><div class="barra ${o.tasa_respuesta >= 0.4 ? 'ok' : ''}">
                  <span style="width:${Math.round((o.tasa_respuesta || 0) * 100)}%"></span></div>
                <div class="small mono">${pct(o.tasa_respuesta)}</div></td>
            <td class="small">${o.calidad_ok} ok · ${o.calidad_sospechosa} sosp. ·
              ${o.calidad_pendiente} pend.</td>
          </tr>`).join('')}</tbody></table></div>`
          : vacio('Este panel todavía no tuvo olas.', '📨')}
      </div>
    </div>

    ${listaPersonas('Sin contacto hace más de 30 días', t.ultimo_contacto.mas_dormidos,
      'Son los que primero se pierden: cuando vuelve la convocatoria, ya no contestan.')}
    ${listaPersonas('Nunca contactados', t.ultimo_contacto.nunca_contactados,
      'Están en el panel y nunca se los convocó.')}
    ${listaPersonas('Más convocados', t.convocatorias.mas_convocados,
      'Los candidatos a fatiga: revisá antes de sumarlos a otra ola.')}`;
}

function barras(distribucion, etiquetas, total) {
  return `<div class="barras">${Object.entries(etiquetas).map(([clave, etiqueta]) => {
    const n = distribucion[clave] || 0;
    const proporcion = total ? n / total : 0;
    return `<div class="fila-barra">
      <span>${esc(etiqueta)}</span>
      <span class="barra"><span style="width:${Math.round(proporcion * 100)}%"></span></span>
      <span class="cifra">${n} · ${pct(proporcion)}</span>
    </div>`;
  }).join('')}</div>`;
}

function listaPersonas(titulo, personas, bajada) {
  if (!personas?.length) return '';
  return `<div class="card">
    <div class="card-header"><span class="card-header-title">${esc(titulo)}</span>
      <span class="small muted">${esc(bajada)}</span></div>
    <div class="card-body tight"><div class="table-wrap"><table>
      <thead><tr><th>Persona</th><th>Convocatorias</th><th>Respuestas</th>
        <th>Tasa</th><th>Último contacto</th></tr></thead>
      <tbody>${personas.map((p) => `<tr data-panelista="${esc(p.id_persona)}"
            style="cursor:pointer">
        <td><div class="td-strong">${esc(p.nombre || '—')}</div>
            <div class="td-muted small">${esc(p.localidad || '')}
              ${p.tramo_etario ? `· ${esc(p.tramo_etario)}` : ''}</div></td>
        <td class="mono">${p.convocatorias}</td>
        <td class="mono">${p.respuestas}</td>
        <td class="mono">${pct(p.tasa_respuesta)}</td>
        <td class="small">${p.ultimo_contacto ? hace(p.ultimo_contacto) : 'nunca'}</td>
      </tr>`).join('')}</tbody>
    </table></div></div>
  </div>`;
}

File: functions/panel_api/participacion.py
"""R2.1 y R2.6 — participación por ola y tablero de salud del panel.

La Fase 1 ya captura el dato: `participacion` tiene una fila por (encuesta,
persona) con cuándo se la convocó, si respondió y en qué estado quedó la
calidad. Lo que faltaba era leerlo. Este módulo no agrega información: la
consolida y la expone.

Tres lecturas, que son las tres preguntas que se le hacen a un panel:

* **¿Responde?** Tasa de respuesta, global y por ola. Es la métrica de
  salud más directa: un panel con 15 % de respuesta no es un panel de mil
  personas, es uno de ciento cincuenta.
* **¿Hace cuánto que no lo molesto?** Tiempo desde el último contacto, por
  persona y en distribución. Sirve para las dos cosas: no quemar a quien
  acaba de responder tres olas, y no dejar dormido a quien no se contactó
  en un año (después no vuelve).
* **¿Reparto parejo?** Distribución de convocatorias. Si el 10 % del panel
  se come el 60 % de las convocatorias, la muestra no es del panel: es de
  ese 10 %.

Todo pasa en la bóveda. Las listas de personas salen con `id_persona` y
nombre, porque el tablero se usa para decidir a quién convocar; quien lo
abre está viendo PII y la ruta registra la reidentificación.
"""

from . import db
from .errores import NoEncontrado

TRAMOS_CONTACTO = (
    ("nunca", None, None),
    ("hasta_30_dias", 0, 30),
    ("31_a_90_dias", 31, 90),
    ("91_a_180_dias", 91, 180),
    ("mas_de_180_dias", 181, None),
)

TRAMOS_CONVOCATORIAS = ("0", "1", "2", "3", "4", "5_o_mas")

DIAS_DORMIDO = 30
"""Desde cuántos días sin contacto una persona entra en la lista de dormidos.

La lista es para actuar, no para mirar: ordenar a todo el panel por
antigüedad de contacto y mostrar los primeros quince no dice nada cuando a
todos se los convocó la semana pasada. Un mes es cuando la falta de contacto
empieza a importar."""


def _tasa(respondieron, convocados):
    return round(respondieron / convocados, 4) if convocados else None


def por_ola(conn, panel_id=None):
    """R2.1 — una fila por encuesta: convocados, respuestas, tasa y calidad."""
    filas = db.todas(
        conn,
        """
        select e.id, e.panel_id, e.nombre, e.fecha_campo, e.estado, e.ref_estudio,
               pa.nombre as panel,
               count(p.id)::int                                          as convocados,
               count(p.id) filter (where p.respondio)::int               as respondieron,
               count(p.id) filter (where p.calidad_estado = 'ok')::int    as calidad_ok,
               count(p.id) filter (where p.calidad_estado = 'sospechoso')::int
                                                                          as calidad_sospechosa,
               count(p.id) filter (where p.calidad_estado = 'pendiente')::int
                                                                          as calidad_pendiente,
               min(p.convocado_en) as primera_convocatoria,
               max(p.respondio_en) as ultima_respuesta
          from encuesta e
          join panel pa on pa.id = e.panel_id
          left join participacion p on p.encuesta_id = e.id
         where (%s::bigint is null or e.panel_id = %s::bigint)
         group by e.id, e.panel_id, e.nombre, e.fecha_campo, e.estado,
                  e.ref_estudio, pa.nombre
         order by e.fecha_campo desc nulls last, e.id desc
        """,
        (panel_id, panel_id),
    )
    return [
        {
            "encuesta_id": f["id"],
            "panel_id": f["panel_id"],
            "panel": f["panel"],
            "nombre": f["nombre"],
            "fecha_campo": f["fecha_campo"].isoformat() if f["fecha_campo"] else None,
            "estado": f["estado"],
            "ref_estudio": str(f["ref_estudio"]),
            "convocados": f["convocados"],
            "respondieron": f["respondieron"],
            "tasa_respuesta": _tasa(f["respondieron"], f["convocados"]),
            "calidad_ok": f["calidad_ok"],
            "calidad_sospechosa": f["calidad_sospechosa"],
            "calidad_pendiente": f["calidad_pendiente"],
            "primera_convocatoria": (
                f["primera_convocatoria"].isoformat() if f["primera_convocatoria"] else None
            ),
            "ultima_respuesta": (
                f["ultima_respuesta"].isoformat() if f["ultima_respuesta"] else None
            ),
        }
        for f in filas
    ]


def _por_persona(conn, panel_id, estado="activo"):
    """Una fila por miembro del panel con su historial de participación.

    El `left join` es lo que hace útil la consulta: quien nunca fue
    convocado tiene que aparecer con 0, y con un `inner join` sería
    invisible. Un miembro nunca convocado es exactamente el caso que el
    tablero está para encontrar.
    """
    return db.todas(
        conn,
        """
        select m.id_persona, p.nombre, p.email, d.sexo, d.localidad, d.tramo_etario,
               count(pa.id)::int                            as convocatorias,
               count(pa.id) filter (where pa.respondio)::int as respuestas,
               max(pa.convocado_en)                         as ultimo_contacto,
               max(pa.respondio_en)                         as ultima_respuesta,
               count(pa.id) filter (where pa.calidad_estado = 'sospechoso')::int
                                                            as calidad_sospechosa
          from membresia m
          join persona p on p.id_persona = m.id_persona
          left join v_demografia d on d.id_persona = m.id_persona
          left join participacion pa on pa.id_persona = m.id_persona
          left join encuesta e on e.id = pa.encuesta_id and e.panel_id = m.panel_id
         where m.panel_id = %s and (%s::text is null or m.estado = %s::text)
           and (pa.id is null or e.id is not null)
         group by m.id_persona, p.nombre, p.email, d.sexo, d.localidad, d.tramo_etario
         order by p.nombre nulls last
        """,
        (panel_id, estado, estado),
    )


def _dias(fecha, ahora):
    if fecha is None:
        return None
    return max(0, (ahora - fecha).days)


def _tramo_contacto(dias):
    if dias is None:
        return "nunca"
    for nombre, desde, hasta in TRAMOS_CONTACTO[1:]:
        if (desde is None or dias >= desde) and (hasta is None or dias <= hasta):
            return nombre
    return "mas_de_180_dias"


def _tramo_convocatorias(cuantas):
    return str(cuantas) if cuantas < 5 else "5_o_mas"


def tablero(conn, panel_id, estado="activo", limite_listas=15):
    """R2.6 — tablero de participación de un panel.

    Devuelve las tres lecturas juntas más dos listas cortas para actuar: a
    quién hace más que no se contacta, y a quién se está convocando de más.
    """
    panel = db.una(
        conn, "select id, nombre from panel where id = %s", (panel_id,)
    )
    if not panel:
        raise NoEncontrado(f"No existe el panel {panel_id}.")

    ahora = db.una(conn, "select now() as ahora")["ahora"]
    filas = _por_persona(conn, panel_id, estado)

    personas = []
    for f in filas:
        dias = _dias(f["ultimo_contacto"], ahora)
        personas.append({
            "id_persona": str(f["id_persona"]),
            "nombre": f["nombre"],
            "email": f["email"],
            "sexo": f["sexo"],
            "localidad": f["localidad"],
            "tramo_etario": f["tramo_etario"],
            "convocatorias": f["convocatorias"],
            "respuestas": f["respuestas"],
            "tasa_respuesta": _tasa(f["respuestas"], f["convocatorias"]),
            "calidad_sospechosa": f["calidad_sospechosa"],
            "ultimo_contacto": (
                f["ultimo_contacto"].isoformat() if f["ultimo_contacto"] else None
            ),
            "dias_sin_contacto": dias,
            "tramo_contacto": _tramo_contacto(dias),
        })

    olas = por_ola(conn, panel_id)
    convocados = sum(o["convocados"] for o in olas)
    respondieron = sum(o["respondieron"] for o in olas)

    distribucion_contacto = {nombre: 0 for nombre, _, _ in TRAMOS_CONTACTO}
    for persona in personas:
        distribucion_contacto[persona["tramo_contacto"]] += 1

    distribucion_convocatorias = {tramo: 0 for tramo in TRAMOS_CONVOCATORIAS}
    for persona in personas:
        distribucion_convocatorias[_tramo_convocatorias(persona["convocatorias"])] += 1

    total_personas = len(personas)
    nunca = distribucion_contacto["nunca"]
    dias_conocidos = [p["dias_sin_contacto"] for p in personas
                      if p["dias_sin_contacto"] is not None]

    # Concentración: qué parte de las convocatorias se come el 10 % más
    # convocado. Es la lectura que dice si la muestra es del panel o de un
    # pedacito del panel.
    por_convocatorias = sorted(
        personas, key=lambda p: (-p["convocatorias"], p["id_persona"])
    )
    cabeza = max(1, round(total_personas * 0.1)) if total_personas else 0
    convocatorias_totales = sum(p["convocatorias"] for p in personas)
    de_la_cabeza = sum(p["convocatorias"] for p in por_convocatorias[:cabeza])

    return {
        "panel_id": panel["id"],
        "panel": panel["nombre"],
        "estado_membresia": estado,
        "miembros": total_personas,
        "olas": olas,
        "respuesta": {
            "convocatorias_emitidas": convocados,
            "respuestas": respondieron,
            "tasa_respuesta": _tasa(respondieron, convocados),
            "miembros_que_respondieron_alguna": sum(
                1 for p in personas if p["respuestas"]
            ),
            "miembros_nunca_convocados": nunca,
        },
        "ultimo_contacto": {
            "distribucion": distribucion_contacto,
            "mediana_dias": (
                sorted(dias_conocidos)[len(dias_conocidos) // 2]
                if dias_conocidos else None
            ),
            "umbral_dormido_dias": DIAS_DORMIDO,
            "mas_dormidos": sorted(
                (
                    p for p in personas
                    if p["dias_sin_contacto"] is not None
                    and p["dias_sin_contacto"] >= DIAS_DORMIDO
                ),
                key=lambda p: -p["dias_sin_contacto"],
            )[:limite_listas],
            "nunca_contactados": [
                p for p in personas if p["dias_sin_contacto"] is None
            ][:limite_listas],
        },
        "convocatorias": {
            "distribucion": distribucion_convocatorias,
            "promedio_por_miembro": (
                round(convocatorias_totales / total_personas, 2)
                if total_personas else None
            ),
            "maximo": max((p["convocatorias"] for p in personas), default=0),
            "concentracion_decil_superior": (
                round(de_la_cabeza / convocatorias_totales, 4)
                if convocatorias_totales else None
            ),
            "mas_convocados": por_convocatorias[:limite_listas],
        },
    }
